chasm/lib/scheduler/scheduler.go

```diff
@@ -0,0 +1,275 @@
+				UpdateTime: timestamppb.New(zero),
+			},
+			InitialPatch:  patch,
+			Namespace:     namespace,
+			NamespaceId:   namespaceID,
+			ScheduleId:    scheduleID,
+			ConflictToken: scheduler.InitialConflictToken,
+		},
+		cacheConflictToken: scheduler.InitialConflictToken,
+		Backfillers:        make(chasm.Map[string, *Backfiller]),
+	}
+
+	invoker := NewInvoker(ctx, sched)
+	generator := NewGenerator(ctx, sched, invoker)
+	sched.Invoker = chasm.NewComponentField(ctx, invoker)
+	sched.Generator = chasm.NewComponentField(ctx, generator)
+
+	return sched
+}
+
+func (s *Scheduler) LifecycleState(ctx chasm.Context) chasm.LifecycleState {
+	if s.Closed {
+		return chasm.LifecycleStateCompleted
+	}
+
+	return chasm.LifecycleStateRunning
+}
+
+// NewRangeBackfiller returns an intialized Backfiller component, which should
+// be parented under a Scheduler root node.
+func (s *Scheduler) NewRangeBackfiller(
+	ctx chasm.MutableContext,
+	request *schedulepb.BackfillRequest,
+) *Backfiller {
+	backfiller := newBackfiller(ctx, s)
+	backfiller.BackfillerInternal.Request = &schedulespb.BackfillerInternal_BackfillRequest{
+		BackfillRequest: request,
+	}
+	s.addBackfiller(ctx, backfiller)
+	return backfiller
+}
+
+// NewImmediateBackfiller returns an intialized Backfiller component, which should
+// be parented under a Scheduler root node.
+func (s *Scheduler) NewImmediateBackfiller(
+	ctx chasm.MutableContext,
+	request *schedulepb.TriggerImmediatelyRequest,
+) *Backfiller {
+	backfiller := newBackfiller(ctx, s)
+	backfiller.BackfillerInternal.Request = &schedulespb.BackfillerInternal_TriggerRequest{
+		TriggerRequest: request,
+	}
+	s.addBackfiller(ctx, backfiller)
+	return backfiller
+}
+
+// addBackfiller adds the backfiller to the scheduler tree, and adds a task to
+// kick off backfill processing.
+func (s *Scheduler) addBackfiller(
+	ctx chasm.MutableContext,
+	backfiller *Backfiller,
+) {
+	s.Backfillers[backfiller.BackfillId] = chasm.NewComponentField(ctx, backfiller)
+	ctx.AddTask(backfiller, chasm.TaskAttributes{}, &schedulespb.BackfillerTask{})
+}
+
+// useScheduledAction returns true when the Scheduler should allow scheduled
+// actions to be taken.
+//
+// When decrement is true, the schedule's state's `RemainingActions` counter is
+// decremented when an action can be taken. When decrement is false, no state
+// is mutated.
+func (s *Scheduler) useScheduledAction(decrement bool) bool {
+	scheduleState := s.Schedule.State
+
+	// If paused, don't do anything.
+	if scheduleState.Paused {
+		return false
+	}
+
+	// If unlimited actions, allow.
+	if !scheduleState.LimitedActions {
+		return true
+	}
+
+	// Otherwise check and decrement limit.
+	if scheduleState.RemainingActions > 0 {
+		if decrement {
+			scheduleState.RemainingActions--
+
+			// The conflict token is updated because a client might be in the process of
+			// preparing an update request that increments their schedule's RemainingActions
+			// field.
+			s.updateConflictToken()
```

Comment:
The conflict token could have been calculated from the scheduler state to avoid having to remember to update it every time, which can lead to subtle bugs.